### PR TITLE
feat(conf/host) Added functionality to update a host

### PR DESCRIPTION
**Pull request by callapa**
## Description

- Fixed some issues due to the Symfony 5 update
- Refactoring to make the code more readable
- Added functionality to update a host

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.10.x
- [ ] 21.04.x
- [ ] 21.10.x
- [x] 22.04.x (master)

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Centreon/Domain/ServiceConfiguration/Interfaces/ServiceConfigurationRepositoryInterface.php [hunk 1]
```diff
@@ -87,4 +87,12 @@ public function findService(int $serviceId): ?Service;
...
      */
     public function findServicesByHost(Host $host): array;
+
+    /**
+     * Removes all services related to a host.
+     *
+     * @param int $hostId
+     * @throws \Exception
+     */
+    public function removeServicesOnhost(int $hostId): void;
```
Comment (a-launois) [inline src/Centreon/Domain/ServiceConfiguration/Interfaces/ServiceConfigurationRepositoryInterface.php:97]:
```suggestion
    public function removeServicesOnHost(int $hostId): void;
```

File: src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php [hunk 2]
```diff
@@ -325,4 +334,31 @@ public function getHostConfigurationService(): HostConfigurationServiceInterface
...
+     * @param Host $host
+     * @throws ServiceConfigurationException
+     * @throws \Assert\AssertionFailedException
+     */
+    public function removeServices(Host $host): void
+    {
+        Assertion::notNull($host->getId(), 'Host::id');
+        try {
+            $this->debug('Remove services from a host', ['host_id' => $host->getId()]);
+            $this->serviceRepository->removeServicesOnhost($host->getId());
```
Comment (a-launois) [inline src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php:348]:
```suggestion
            $this->serviceRepository->removeServicesOnHost($host->getId());
```

File: src/Centreon/Infrastructure/ServiceConfiguration/ServiceConfigurationRepositoryRDB.php [hunk 3]
```diff
@@ -373,4 +373,19 @@ public function findServicesByHost(Host $host): array
         }
         return $services;
     }
+
+    /**
+     * @inheritDoc
+     */
+    public function removeServicesOnhost(int $hostId): void
```
Comment (a-launois) [inline src/Centreon/Infrastructure/ServiceConfiguration/ServiceConfigurationRepositoryRDB.php:380]:
```suggestion
    public function removeServicesOnHost(int $hostId): void
```

File: src/Centreon/Domain/HostConfiguration/HostConfigurationService.php [hunk 4]
```diff
@@ -70,6 +74,10 @@ class HostConfigurationService implements HostConfigurationServiceInterface
      */
     private $hostGroupService;
 
+    /**
+     * @var ContactInterface
+     */
+    private $contact;
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/HostConfiguration/HostConfigurationService.php:80]:
```suggestion
    private $contact;

```

File: src/Centreon/Domain/HostConfiguration/HostConfigurationService.php [hunk 5]
```diff
@@ -95,144 +105,112 @@ public function __construct(
...
+            $this->hostConfigurationRepository->updateHost($host);
+
+            if ($transactionAlreadyStarted === false) {
+                $this->dataStorageEngine->commitTransaction();
+            }
+        } catch (\Throwable $ex) {
+            if ($transactionAlreadyStarted === false) {
+                $this->dataStorageEngine->rollbackTransaction();
+            }
+            throw HostConfigurationServiceException::errorOnUpdatingAHost($ex);
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/HostConfiguration/HostConfigurationService.php:213]: Could you add debug logs on transaction  as you did in the `addHost` method

File: src/Centreon/Domain/HostConfiguration/HostConfigurationService.php [hunk 6]
```diff
@@ -504,4 +489,104 @@ private function createHostGroupsBeforeLinking(array $groups): void
...
+        if ($engineConfiguration === null) {
+            throw new HostConfigurationException(_('Unable to find the Engine configuration'));
+        }
+        $this->debug(
+            'Engine configuration',
+            ['id' => $engineConfiguration->getId(), 'name' => $engineConfiguration->getName()]
+        );
+        $safedHostName = $engineConfiguration->removeIllegalCharacters($host->getName());
+        if (empty($safedHostName)) {
+            throw new HostConfigurationException(_('Host name can not be empty'));
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/HostConfiguration/HostConfigurationService.php:515]:
```suggestion
            throw HostConfigurationServiceException::hostNameCanNotBeEmpty();
```

File: src/Centreon/Domain/HostConfiguration/HostConfigurationService.php [hunk 7]
```diff
@@ -504,4 +489,104 @@ private function createHostGroupsBeforeLinking(array $groups): void
...
+     * @throws \Centreon\Domain\Engine\EngineException
+     */
+    private function checkIllegalCharactersInHostName(Host $host): void
+    {
+        $this->info('Check illegal characters in host name');
+        $engineConfiguration = $this->engineConfigurationService->findEngineConfigurationByName(
+            $host->getMonitoringServer()->getName()
+        );
+        if ($engineConfiguration === null) {
+            throw new HostConfigurationException(_('Unable to find the Engine configuration'));
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/HostConfiguration/HostConfigurationService.php:507]: Shouldn't you create a static method in HostConfigurationServiceException for this one ?

File: src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php [hunk 8]
```diff
@@ -325,4 +334,31 @@ public function getHostConfigurationService(): HostConfigurationServiceInterface
...
+                    _('Error on removing services from the host #%d (Reason: %s)'),
+                    $host->getId(),
+                    $ex->getMessage()
+                )
+            );
+            throw new ServiceConfigurationException(
+                sprintf(_('Error on removing services from the host #%d'), $host->getId()),
+                0,
+                $ex
+            );
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php:361]: Can you create a static method in `ServiceConfigurationException` to handle this Exception ?

**Review by jeremyjaouen — APPROVED** [review]

**Review by a-launois — APPROVED** [review]

**Pull request merged**